flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java

```diff
@@ -0,0 +1,264 @@
+        } else {
+            context.goToFailing(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getFailureCause());
+        }
+    }
+
+    CompletableFuture<String> getOperationCompletionFuture() {
+        return stopWithSavepointOperationManager.getSavepointPathFuture();
+    }
+
+    interface Context extends StateWithExecutionGraph.Context {
+        /**
+         * Asks how to handle the failure.
+         *
+         * @param failure failure describing the failure cause
+         * @return {@link Executing.FailureResult} which describes how to handle the failure
+         */
+        Executing.FailureResult howToHandleFailure(Throwable failure);
+
+        /**
+         * Transitions into the {@link Canceling} state.
+         *
+         * @param executionGraph executionGraph to pass to the {@link Canceling} state
+         * @param executionGraphHandler executionGraphHandler to pass to the {@link Canceling} state
+         * @param operatorCoordinatorHandler operatorCoordinatorHandler to pass to the {@link
+         *     Canceling} state
+         */
+        void goToCanceling(
+                ExecutionGraph executionGraph,
+                ExecutionGraphHandler executionGraphHandler,
+                OperatorCoordinatorHandler operatorCoordinatorHandler);
+
+        /**
+         * Transitions into the {@link Restarting} state.
+         *
+         * @param executionGraph executionGraph to pass to the {@link Restarting} state
+         * @param executionGraphHandler executionGraphHandler to pass to the {@link Restarting}
+         *     state
+         * @param operatorCoordinatorHandler operatorCoordinatorHandler to pas to the {@link
+         *     Restarting} state
+         * @param backoffTime backoffTime to wait before transitioning to the {@link Restarting}
+         *     state
+         */
+        void goToRestarting(
+                ExecutionGraph executionGraph,
+                ExecutionGraphHandler executionGraphHandler,
+                OperatorCoordinatorHandler operatorCoordinatorHandler,
+                Duration backoffTime);
+
+        /**
+         * Transitions into the {@link Failing} state.
+         *
+         * @param executionGraph executionGraph to pass to the {@link Failing} state
+         * @param executionGraphHandler executionGraphHandler to pass to the {@link Failing} state
+         * @param operatorCoordinatorHandler operatorCoordinatorHandler to pass to the {@link
+         *     Failing} state
+         * @param failureCause failureCause describing why the job execution failed
+         */
+        void goToFailing(
+                ExecutionGraph executionGraph,
+                ExecutionGraphHandler executionGraphHandler,
+                OperatorCoordinatorHandler operatorCoordinatorHandler,
+                Throwable failureCause);
```

Comment:
How are we going back to `Executing` if the savepoint future failed? Or can this not happen?

Reply:
My assumption while writing the code was that a savepoint failure leads to a transition of the Task from RUNNING to FAILED, hence we get a notification via updateTaskExecutionState about this failure, where we then go into the Failing state.

However, I see that there might be a scenario where the Checkpoint coordinator notifies us (via the savepoint future) that the savepoint failed, before we receive the notification from the TaskExecutor that the Task failed. In this case, we would indeed be stuck in the StopWithSavepoint state, somehow depending on the Task failure to get us out there.
Also, if the implementation of the Task changes, we will have to touch the StopWithSavepoint state.

I'm probably a bit too test test driven in my development ;)